phoenix-core-client/src/main/java/org/apache/phoenix/expression/util/bson/CommonComparisonExpressionUtils.java

```diff
@@ -0,0 +1,271 @@
+   * @return If the field key exists in the document, return the corresponding value. Else
+   * return null.
+   */
+  public static BsonValue getNestedFieldVal(BsonValue value, int idx,
+                                            final String documentFieldKey) {
+    if (idx == documentFieldKey.length()) {
+      return value;
+    }
+    int curIdx = idx;
+    if (documentFieldKey.charAt(curIdx) == '.') {
+      BsonDocument nestedMap = value != null && value.isDocument() ? (BsonDocument) value : null;
+      if (nestedMap == null) {
+        LOGGER.error("Incorrect access. Should have found nested map for value: {}", value);
+        return null;
+      }
+      curIdx++;
+      StringBuilder sb = new StringBuilder();
+      for (; curIdx < documentFieldKey.length(); curIdx++) {
+        if (documentFieldKey.charAt(curIdx) == '.' || documentFieldKey.charAt(curIdx) == '[') {
+          BsonValue nestedValue = nestedMap.get(sb.toString());
+          if (nestedValue == null) {
+            return null;
+          }
+          return getNestedFieldVal(nestedValue, curIdx, documentFieldKey);
+        } else {
+          sb.append(documentFieldKey.charAt(curIdx));
+        }
+      }
+      return nestedMap.get(sb.toString());
+    } else if (documentFieldKey.charAt(curIdx) == '[') {
+      curIdx++;
+      StringBuilder arrayIdxStr = new StringBuilder();
+      while (documentFieldKey.charAt(curIdx) != ']') {
+        arrayIdxStr.append(documentFieldKey.charAt(curIdx));
+        curIdx++;
+      }
+      curIdx++;
+      int arrayIdx = Integer.parseInt(arrayIdxStr.toString());
+      BsonArray nestedList = value != null && value.isArray() ? (BsonArray) value : null;
+      if (nestedList == null) {
+        LOGGER.error("Incorrect access. Should have found nested list for value: {}", value);
+        return null;
+      }
+      if (arrayIdx >= nestedList.size()) {
+        LOGGER.error(
+            "Incorrect access. Nested list size {} is less than attempted index access at {}",
+            nestedList.size(), arrayIdx);
+        return null;
+      }
+      BsonValue valueAtIdx = nestedList.get(arrayIdx);
+      if (curIdx == documentFieldKey.length()) {
+        return valueAtIdx;
+      }
+      return getNestedFieldVal(valueAtIdx, curIdx, documentFieldKey);
+    }
+    LOGGER.error("This is erroneous case. getNestedFieldVal should not be used for "
+        + "top level document fields");
+    return null;
+  }
+
+  /**
+   * Compares two values: value and compareValue. If both have same data type and the value of
+   * actual value is less than the compareValue, then returns true. Otherwise, returns false.
+   *
+   * @param value The actual value to compare against compareValue.
+   * @param compareValue The value against which the actual value is compared.
+   * @return True if both have same data type and the value is less than compareValue.
+   */
+  public static boolean lessThan(BsonValue value, BsonValue compareValue) {
```

Comment:
We can combine the following four methods into one compare by providing the comparison operator.

Reply:
done